tensorflow/java/src/main/java/org/tensorflow/Tensor.java

```diff
@@ -28,89 +28,123 @@
  *
  * <p><b>WARNING:</b> Resources consumed by the Tensor object <b>must</b> be explicitly freed by
  * invoking the {@link #close()} method when the object is no longer needed. For example, using a
- * try-with-resources block like:
+ * try-with-resources block:
  *
  * <pre>{@code
- * try(Tensor t = Tensor.create(...)) {
+ * try (Tensor t = Tensor.create(...)) {
  *   doSomethingWith(t);
  * }
  * }</pre>
  */
-public final class Tensor implements AutoCloseable {
+public final class Tensor<T> implements AutoCloseable {
 
   /**
-   * Create a Tensor from a Java object.
+   * Creates a Tensor from a Java object.
    *
-   * <p>A Tensor is a multi-dimensional array of elements of a limited set of types ({@link
-   * DataType}). Thus, not all Java objects can be converted to a Tensor. In particular, {@code obj}
-   * must be either a primitive (float, double, int, long, boolean) or a multi-dimensional array of
-   * one of those primitives. For example:
+   * <p>A {@code Tensor} is a multi-dimensional array of elements of a limited set of types ({@link
+   * types}), so not all Java objects can be converted to a {@code Tensor}. In particular, the
+   * argument {@code obj} must be either a primitive (float, double, int, long, boolean, byte) or a
+   * multi-dimensional array of one of those primitives. The argument {@code type} specifies how to
+   * interpret the first argument as a TensorFlow type. For example:
    *
    * <pre>{@code
    * // Valid: A 64-bit integer scalar.
-   * Tensor s = Tensor.create(42L);
+   * Tensor<Long> s = Tensor.create(42L, Long.class);
    *
    * // Valid: A 3x2 matrix of floats.
    * float[][] matrix = new float[3][2];
-   * Tensor m = Tensor.create(matrix);
+   * Tensor<Float> m = Tensor.create(matrix, Float.class);
    *
    * // Invalid: Will throw an IllegalArgumentException as an arbitrary Object
    * // does not fit into the TensorFlow type system.
-   * Tensor o = Tensor.create(new Object());
+   * Tensor<?> o = Tensor.create(new Object(), ...);
    *
    * // Invalid: Will throw an IllegalArgumentException since there are
    * // a differing number of elements in each row of this 2-D array.
    * int[][] twoD = new int[2][];
    * twoD[0] = new int[1];
    * twoD[1] = new int[2];
-   * Tensor x = Tensor.create(twoD);
+   * Tensor<Integer> x = Tensor.create(twoD, Integer.class);
    * }</pre>
    *
-   * {@link DataType#STRING} typed Tensors are multi-dimensionary arrays of arbitrary byte sequences
-   * and thus have {@code byte[]} and not {@code String}-valued elements. For example:
+   * {@link types.String} typed Tensors are multi-dimensional arrays of arbitrary byte sequences, so
+   * can be initialized from arrays of {@code byte[]} elements. For example:
    *
    * <pre>{@code
-   * // Valid: A DataType.STRING tensor.
-   * Tensor s = Tensor.create(new byte[]{1, 2, 3});
+   * // Valid: A String tensor.
+   * Tensor<String> s = Tensor.create(new byte[]{1, 2, 3}, String.class);
```

Comment:
I see what you mean about potential confusion, but I think programmers will be more comfortable with using String than a new Bytes class. I think programmers will be mostly creating string Tensors using the Tensors class, which for reasons of both convenience and efficiency ought to allow creating tensors from (arrays of) Java Strings with a specified encoding.

Reply:
Also, calling it a Tensor<String> is actually more compatible with the way tensors are presented in the other language APIs.

Reply:
I agree that it's "really" a bunch of bytes, but programmers will think of it as containing strings. The fact that there is an encoding to and from bytes will be largely a detail, since people are mostly not going to write code manipulating these strings at the byte level anyway. I would even suggest picking a default encoding -- UTF-8 seems to have become
 dominant -- and allowing alternate encodings to be provided.

Reply:
Many of the TensorFlow operations that accept TF_STRING tensors truly expect a sequence of bytes, such as an encoded protocol buffer.  A popular one is [`tf.parse_example`](https://www.tensorflow.org/api_docs/python/tf/parse_example). The serialized protocol buffers cannot be represented as a Java string.

So I guess I'm disputing the claim that "mostly not going to write code manipulating these strings at the byte level anyway"  :).  

Hence, I'm tempted to introduce a `Bytes` type. If later, we decide to add a `String` which implies UTF-8 encoding to `Bytes` as a convenience, maybe that will be okay. But I feel we should be true to what the type really represents?

Reply:
I guess I'm looking at the Python API code and seeing that it's really nice to be able to write things like `tf.constant("filename.txt")` rather than some awkward thing that explicitly encodes the string into an array of bytes. The Python programmer gets to treat Python strings as being the things in tensors, so it seems like a shame to give the Java programmer something clunkier and more foreign when Java strings are arguably more capable than Python strings.

Not to pick nits, but it seems to me that serialized protocol buffers actually can be represented as a Java string, precisely because Java characters are bigger than 8 bits. If encoded or decoded using ISO-8859-1, there will be a one-to-one correspondence between the Java characters and the bytes in the serialized protobuf. Yes, Java strings can have characters that can't be represented as bytes, but I guess I'm not seeing the harm that comes from nevertheless saying these are tensors of String.

Reply:
Of course, there is no real requirement that the type parameter actually correspond to what is stored in the tensor. But the thrust of the recent set of changes was to move away from inventing new types like TFString/Bytes and try to use existing, familiar Java types. I thought you had a good idea there!

Reply:
Taking a step back, I think we agree on wanting to make it convenient to convert Java `String`s into `Tensor`s. Mostly via methods in the (yet to come) `Tensors` class with factory functions.

The point of debate here is what these functions should return, i.e.:

```java
public class Tensors {
  public static Tensor<String> create(String value);  // UTF-8 encoded
  public static Tensor<String> create(String value, Charset charset);
  public static Tensor<String> create(byte[] value);
}
```

*or*

```java
public class Tensors {
  public static Tensor<Bytes> create(String value);  // UTF-8 encoded
  public static Tensor<Bytes> create(String value, Charset charset);
  public static Tensor<Bytes> create(byte[] value);
}
```

My contention was that the latter is preferable. It makes it clear that the element type of the tensor is a sequence of bytes, not a `String`. I feel that this is more accurate and if the developer blinks, it's because this is a distinction worth knowing about.  For example, [`tf.equal`](https://www.tensorflow.org/api_docs/python/tf/equal) on two scalar `Tensor<String>` objects will return `false` even if they are the same "strings" because the encoding use to create the `Tensor<String>` might have been different.

Point taken about the serialized protocol buffers and ISO-8859-1 encoding. But it seems to be more natural to represent the serialized buffer as a `byte[]` rather than a `String`, which is why presumably the [protocol buffer APIs use them](https://developers.google.com/protocol-buffers/docs/javatutorial#parsing-and-serialization).

You example on `tf.constant("filename.txt")` remains equally convenient in both, no? It's just whether the returned value is a `Tensor<Bytes>` or a `Tensor<String>`.

Thoughts? :)

Reply:
You're absolutely right, we can offer exactly the same API whether we call it `Tensor<String>` or `Tensor<Bytes>`. If we go the second route, I would prefer to call it `Tensor<TFString>` rather than `Tensor<Bytes>` because 1) that makes the correspondence to the other language APIs (e.g., `tf.string`) more direct and obvious, 2) it communicates that this is the way you solve the common problem of putting strings into tensors, and 3) still adds the developer eye-blink you're after.

Personally, I think that people will, in the odd case that it matters, still make the same silly encoding/decoding mistakes regardless of what we call the type parameter, so I feel the familiarity of `String` is more valuable than the warning conveyed by the unexpected names `Bytes`/`TFString`.

Reply:
Let's agree to disagree here (regarding familiarity with `String`) :). My fear is that developers will _think_ they're comfortable with `String` and 95% of the time it will work out fine, but those 5% where it will differ from the other types (e.g., `Integer.equals()` and `tf.equals(Tensor<Integer>, Tensor<Integer>)` will be consistent, while `String.equals()` and `tf.equals(Tensor<String>, Tensor<String>)` will not) will be annoying.

That said, other than this `String` business we do agree and there is much to this PR and I'd like to get this in before 1.4 is cut, so let's proceed with the PR as is and we can consider changing `String` to `Bytes` or `TFString` in a follow up (I'll try to find some time to gather more opinions).